applications/StructuralMechanicsApplication/tests/cpp_tests/test_solid_static_sensitivity.cpp

```diff
@@ -0,0 +1,208 @@
+    params.AddEmptyValue("traced_node_id");
+    params["traced_node_id"].SetInt(ResponseNodeId);
+    return Kratos::make_shared<AdjointNodalDisplacementResponseFunction>(*pModelPart, params);
+}
+
+PrimalTestSolver::PrimalTestSolver(ModelPart* pPrimalModelPart, unsigned ResponseNodeId)
+    : mpPrimalModelPart(pPrimalModelPart), mResponseNodeId(ResponseNodeId){};
+
+double PrimalTestSolver::CalculateResponseValue()
+{
+    auto p_solver = CreateSolvingStrategy();
+    p_solver->Initialize();
+    p_solver->Solve();
+    auto p_response_function = ResponseFunctionFactory(mpPrimalModelPart, mResponseNodeId);
+    return p_response_function->CalculateValue(*mpPrimalModelPart);
+}
+
+double PrimalTestSolver::CalculateResponseValue(const unsigned NodeToPerturb,
+                                                const char Direction,
+                                                const double Perturbation)
+{
+    KRATOS_ERROR_IF(Direction != 'x' && Direction != 'y')
+        << "invalid direction: '" << Direction << "'";
+    KRATOS_ERROR_IF(Perturbation <= 0.) << "invalid perturbation: " << Perturbation;
+    const unsigned i_dir = (Direction == 'x') ? 0 : 1;
+    auto& r_node = mpPrimalModelPart->GetNode(NodeToPerturb);
+    r_node.GetInitialPosition()[i_dir] += Perturbation;
+    const double response_value = CalculateResponseValue();
+    r_node.GetInitialPosition()[i_dir] -= Perturbation;
+    return response_value;
+}
+
+SolvingStrategyType::Pointer PrimalTestSolver::CreateSolvingStrategy()
+{
+    LinearSolverType::Pointer p_linear_solver =
+        Kratos::make_shared<SkylineLUCustomScalarSolver<SparseSpaceType, LocalSpaceType>>();
+    SchemeType::Pointer p_scheme =
+        Kratos::make_shared<ResidualBasedIncrementalUpdateStaticScheme<SparseSpaceType, LocalSpaceType>>();
+    ConvergenceCriteriaType::Pointer p_conv_criteria =
+        Kratos::make_shared<ResidualCriteria<SparseSpaceType, LocalSpaceType>>(1e-12, 1e-14);
+    return Kratos::make_shared<ResidualBasedNewtonRaphsonStrategy<SparseSpaceType, LocalSpaceType, LinearSolverType>>(
+        *mpPrimalModelPart, p_scheme, p_linear_solver, p_conv_criteria, 30,
+        true, false, true);
+}
+
+AdjointTestSolver::AdjointTestSolver(ModelPart* pAdjointModelPart, unsigned ResponseNodeId)
+    : mpAdjointModelPart(pAdjointModelPart), mResponseNodeId(ResponseNodeId)
+{
+    auto p_adjoint_response_function =
+        ResponseFunctionFactory(mpAdjointModelPart, mResponseNodeId);
+    auto p_adjoint_solver = CreateAdjointSolvingStrategy(p_adjoint_response_function);
+    p_adjoint_solver->Initialize();
+    p_adjoint_solver->Solve();
+    SensitivityBuilder sensitivity_builder(
+        Parameters(R"({"nodal_solution_step_sensitivity_variables": ["SHAPE_SENSITIVITY"], "build_mode": "static"})"),
+        *mpAdjointModelPart, p_adjoint_response_function);
+    sensitivity_builder.Initialize();
+    sensitivity_builder.UpdateSensitivities();
```

Comment:
Cosmetics: The solution happening inside the constructor of the AdjointTestSolver seems weird. Especially because that's not the case for the primal solver. Up to you though.

Reply:
It is the most efficient for testing. The alternative is to solve the primal each time we want to solve the adjoint.